destination/writer.go

```diff
@@ -0,0 +1,164 @@
+
+package destination
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/data/azcosmos"
+	"github.com/conduitio-labs/conduit-connector-cosmos-nosql/common"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+)
+
+// Writer implements a writer logic for the Neo4j Destination.
+type writer struct {
+	containerClient *azcosmos.ContainerClient
+	partitionKey    azcosmos.PartitionKey
+}
+
+// New creates a new instance of the [Writer].
+func newWriter(config Config) (*writer, error) {
+	w := &writer{
+		partitionKey: azcosmos.NewPartitionKeyString(config.PartitionValue),
+	}
+
+	// create an KeyCredential containing the account's primary key
+	cred, err := azcosmos.NewKeyCredential(config.PrimaryKey)
+	if err != nil {
+		return nil, fmt.Errorf("new key credential: %w", err)
+	}
+
+	// create a new instance of Cosmos client
+	client, err := azcosmos.NewClientWithKey(config.URI, cred, nil)
+	if err != nil {
+		return nil, fmt.Errorf("new client: %w", err)
+	}
+
+	// create a new instance of the specific container client
+	w.containerClient, err = client.NewContainer(config.Database, config.Container)
+	if err != nil {
+		return nil, fmt.Errorf("new container client: %w", err)
+	}
+
+	return w, nil
+}
+
+// Write writes a record to the destination.
+func (w *writer) Write(ctx context.Context, record sdk.Record) error {
+	err := sdk.Util.Destination.Route(ctx, record,
+		w.create,
+		w.replace,
+		w.delete,
+		w.create,
+	)
+	if err != nil {
+		return fmt.Errorf("route record: %w", err)
+	}
+
+	return nil
+}
+
+func (w *writer) create(ctx context.Context, record sdk.Record) error {
+	payload := make(sdk.StructuredData)
+	if err := json.Unmarshal(record.Payload.After.Bytes(), &payload); err != nil {
+		return fmt.Errorf("unmarshal payload: %w", err)
+	}
+
+	// if sdk.Record.Key does not contain the 'id' key
+	if _, okPayload := payload[common.KeyID]; !okPayload {
+		key := make(sdk.StructuredData)
+		if err := json.Unmarshal(record.Key.Bytes(), &key); err != nil {
+			return fmt.Errorf("unmarshal key: %w", err)
+		}
+
+		id, okKey := key[common.KeyID]
+		if !okKey {
+			return fmt.Errorf("neither the sdk.Record.Key nor the sdk.Record.Payload contains the required %q key",
+				common.KeyID)
+		}
+
+		//nolint:forcetypeassert // there is no point in checking the type because it was unmarshalled above
+		record.Payload.After.(sdk.StructuredData)[common.KeyID] = id.(string)
+	}
+
+	_, err := w.containerClient.CreateItem(ctx, w.partitionKey, record.Payload.After.Bytes(), nil)
+	if err != nil {
+		return fmt.Errorf("create item: %w", err)
+	}
+
+	return nil
+}
+
+func (w *writer) replace(ctx context.Context, record sdk.Record) error {
+	payload := make(sdk.StructuredData)
+	if err := json.Unmarshal(record.Payload.After.Bytes(), &payload); err != nil {
+		return fmt.Errorf("unmarshal payload: %w", err)
+	}
+
+	// if sdk.Record.Key does not contain the 'id' key
+	if _, okPayload := payload[common.KeyID]; !okPayload {
```

Comment:
maybe lets move this code to separate function

Reply:
I've moved this logic to another method 👍